autosummary tables include module attributes

dfremont opened the pull request:
Subject: Adds module attributes to autosummary tables

### Feature or Bugfix
<!-- please choose -->
- Feature

### Purpose

As pointed out in #4058, autosummary currently does not expose module attributes (called "module-level variables" there). Even if they are documented using doc-comments, such attributes do not get mentioned in the generated summary tables. For example, if you use autosummary on this module:

```
#: A module attribute which should be mentioned in the summary
my_attr = 3.14

def func():
    """A function which is actually mentioned in the summary"""
    pass
```

then the generated page documents both `my_attr` and `func`, but only lists `func` in the summary tables. I work on a library which uses module attributes as part of the public API, and would like to make them prominent by listing them in the summary tables for the module; this small PR makes that possible.

### Detail

The PR makes three changes:

- Defines the `attributes` and `all_attributes` template variables (already used for classes) when autosummarizing modules. The `all_attributes` variable lists all module attributes with doc-comments; the `attributes` variable lists the subset of these whose name does not start with an underscore.
- Updates the autosummary documentation to reflect the above (I don't document `all_attributes` since the `all_` versions of all such template variables are undocumented and I'm not sure whether they're supposed to be).
- Adds a new section to the autosummary module template called "Module Attributes", which lists the elements of `attributes` in the same way as currently done for `functions`, etc.

 Some notes:

- I'm not familiar with Sphinx, so it's possible the way I'm searching for the documented module attributes is wasteful (in particular I create a `ModuleAnalyzer` since I couldn't figure out how to access an existing one). Please feel free to completely change my implementation: I would like Sphinx to support this feature, but don't care how it's done.
- If you would rather not have this PR affect existing code, you can remove the change to the module template. Then people will still have the option to make a summary table of module attributes, even if one isn't included by default.
- I'm afraid I don't understand Sphinx enough to write an automated test for this PR - sorry! It does work correctly with my code base, at least.

dfremont committed:
autosummary includes module attributes

dfremont committed:
updated corresponding documentation

dfremont commented:
Oh dear, it looks like a test is failing due to a docstring being processed twice: presumably this is related to the issue I mentioned about creating a new `ModuleAnalyzer`. If someone could give me a hint about how to properly orchestrate this I can try to fix it.

tk0miya commented:
>Oh dear, it looks like a test is failing due to a docstring being processed twice

Finally, I found the reason of this behavior. I just post a fix #7580 now. I'll self-review and merge it tomorrow. Please wait a moment.

tk0miya commented on sphinx/ext/autosummary/generate.py:
```diff
@@ -197,6 +197,21 @@ def get_members(obj: Any, types: Set[str], include_public: List[str] = [],
+
+        # Find module attributes with docstrings
+        attrs, public = [], []
+        from sphinx.pycode import ModuleAnalyzer, PycodeError
```
Could you move this line to the top of this module, please? Basically we go importing there if no reason.

tk0miya commented on sphinx/ext/autosummary/templates/autosummary/module.rst:
```diff
@@ -2,6 +2,17 @@
+
+   .. autosummary::
+   {% for item in attributes %}
+      {{item}}
```
Please add spaces between a template variable.

tk0miya commented on sphinx/ext/autosummary/generate.py:
```diff
@@ -197,6 +197,21 @@ def get_members(obj: Any, types: Set[str], include_public: List[str] = [],
+                    if not attr_name.startswith('_'):
+                        public.append(attr_name)
+        except PycodeError as err:
+            pass
```
Please make a small internal function like `get_members()`. I think it makes this function a bit readable.

tk0miya commented on sphinx/ext/autosummary/generate.py:
```diff
@@ -197,6 +197,21 @@ def get_members(obj: Any, types: Set[str], include_public: List[str] = [],
+                        public.append(attr_name)
+        except PycodeError as err:
+            pass
+        ns['attributes'], ns['all_attributes'] = attrs, public
```
Please add these members to `ns['members']` too. It should contain all of the members including attributes.

tk0miya replied:
This is my wrong. withdrawing.

tk0miya commented on doc/usage/extensions/autosummary.rst:
```diff
@@ -260,8 +260,8 @@ The following variables available in the templates:
+   List containing names of "public" attributes in the class/module.  Only
+   available for classes and modules.
 
 
```
Please add here the changes in this PR. For example:
```
.. versionchanged:: 3.1

   Attributes of modules are supported.
```

tk0miya reviewed (changes requested):
LGTM with nits.

To add a test case, please do the following steps:

1. add a module variable to `tests/roots/test-ext-autosummary/autosummary_dummy_module.py`
2. update a testcase `test_autosummary_generate()` in `tests/test_ext_autosummary.py` to confirm that the stub file will be generated expectedly


Thanks,

tk0miya commented:
Sorry, I closed this PR by my mistake. So reopened again.

tk0miya commented:
> Finally, I found the reason of this behavior. I just post a fix #7580 now. I'll self-review and merge it tomorrow. Please wait a moment.

I merged the fix to 3.x branch. Please merge it to your branch.

tk0miya commented on sphinx/ext/autosummary/generate.py:
```diff
@@ -197,6 +197,21 @@ def get_members(obj: Any, types: Set[str], include_public: List[str] = [],
+            analyzer = ModuleAnalyzer.for_module(name)
+            attr_docs = analyzer.find_attr_docs()
+            for _, attr_name in attr_docs:
+                if attr_name in ns['members']:
```
This is not a good way to filter the documented attributes. The key of `attr_docs` is pairs of namespace and name. The namespace is a parent name of the variable. The namespace of global attributes is an empty string. This code should ensure empty. If none, it would pick up under class object.

dfremont committed:
Merge remote-tracking branch 'sphinx/3.x' into 3.x

dfremont committed:
revisions per comments from tk0miya

dfremont committed:
got public/all attrs backwards

dfremont commented:
Thanks for the detailed comments and help writing the test! I think I've adjusted everything you pointed out above (and fixed a bug where I got the lists of public and all attributes backwards). Let me know if I should change anything else.

tk0miya commented on doc/usage/extensions/autosummary.rst:
```diff
@@ -285,8 +285,12 @@ The following variables available in the templates:
+
+    .. versionchanged:: 3.1
+
+    Attributes of modules are supported.
```
This content should be indented inside the `versionchanged` directive. But it's minor bug. So I'll work this after merging.

tk0miya reviewed (approved):
Thank you for update. LGTM!
